server/server.go

```diff
@@ -1147,3 +1156,188 @@ func (s *Server) getClientConnectURLs() []string {
 
 	return urls
 }
+
+// if the ip is not specified, attempt to resolve it
+func resolveHostPorts(addr net.Listener) []string {
+	hostPorts := make([]string, 0)
+	hp := addr.Addr().(*net.TCPAddr)
+	port := strconv.Itoa(hp.Port)
+	if hp.IP.IsUnspecified() {
+		var ip net.IP
+		ifaces, _ := net.Interfaces()
+		for _, i := range ifaces {
+			addrs, _ := i.Addrs()
+			for _, addr := range addrs {
+				switch v := addr.(type) {
+				case *net.IPNet:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				case *net.IPAddr:
+					ip = v.IP
+					hostPorts = append(hostPorts, net.JoinHostPort(ip.String(), port))
+				default:
+					continue
+				}
+			}
+		}
+	} else {
+		hostPorts = append(hostPorts, net.JoinHostPort(hp.IP.String(), port))
+	}
+	return hostPorts
+}
+
+// format the address of a net.Listener with a protocol
+func formatURL(protocol string, addr net.Listener) []string {
+	hostports := resolveHostPorts(addr)
+	for i, hp := range hostports {
+		hostports[i] = fmt.Sprintf("%s://%s", protocol, hp)
+	}
+	return hostports
+}
+
+// Ports describes URLs that the server can be contacted in
+type Ports struct {
+	Nats       []string `json:"nats,omitempty"`
+	Monitoring []string `json:"monitoring,omitempty"`
+	Cluster    []string `json:"cluster,omitempty"`
+	Profile    []string `json:"profile,omitempty"`
+}
+
+// Returns a Ports struct describing ports where the server can be contacted
+func (s *Server) PortsInfo(opts *Options) Ports {
+	s.mu.Lock()
+	info := s.copyInfo()
+	listener := s.listener
+	httpListener := s.http
+	clusterListener := s.routeListener
+	profileListener := s.profiler
+	s.mu.Unlock()
+
+	ports := Ports{}
+
+	if listener != nil {
+		natsProto := "nats"
+		if info.TLSRequired {
+			natsProto = "tls"
+		}
+		ports.Nats = formatURL(natsProto, listener)
+	}
+
+	if httpListener != nil {
+		monProto := "http"
+		if opts.HTTPSPort != 0 {
+			monProto = "https"
+		}
+		ports.Monitoring = formatURL(monProto, httpListener)
+	}
+
+	if clusterListener != nil {
+		clusterProto := "nats"
+		if opts.Cluster.TLSConfig != nil {
+			clusterProto = "tls"
+		}
+		ports.Cluster = formatURL(clusterProto, clusterListener)
+	}
+
+	if profileListener != nil {
+		ports.Profile = formatURL("http", profileListener)
+	}
+
+	return ports
+}
+
+// Returns the portsFile. If a non-empty dirHint is provided, the dirHint
+// path is used instead of the server option value
+func (s *Server) PortsFile(dirHint string) string {
```

Comment:
Would not pass that as an option to the public call. The public call should return the name of the file based on server's options.

Reply:
It cannot - in the case of the reload, the options are changed, and it cannot be deduced from that.

Reply:
If there is a reload, the new options are set during the Reload() so if you then call PortsFile() that function could find the new file without a need to pass it as a param. I think I don't understand what you are trying to do then.

Reply:
But that's my point, you don't need. If you call deletePorts(p.oldValue) that is enough for this function to construct the file name with the given directory. You don't need PortFile(), or in other words, PortFile is simply a file path Join of the exe name and pid...